From da5d5a6a77c131fb74f318116f0d9c21658a266e Mon Sep 17 00:00:00 2001
From: "psathiyendran@maxlinear.com" <psathiyendran@sgb017.sgsw.maxlinear.com>
Date: Wed, 4 Jun 2025 18:18:31 +0800
Subject: [PATCH] WLANRTSYS-87691 [UPDK] [Wave700] 6GHz Network: Connection
 Issues with Hidden SSIDs on VAP

Issue : In 6GHz while making both Tx and Non Tx SSID as hidden
and try to connect Non-Tx SSID, connection is failing

Root-Cause : While MBSSID is enabled, the probe request carries Non-Tx
SSID IE in Tx BSS, but probe response we will be sending Tx SSID IE,
and also AP hides Non-Tx SSID in MBSSID fieled.

Fix : In Probe-Response hide the Tx SSID in SSID IE, and show only
Non-Tx ssid in MBSSID field for which Probe-Request is sent. This is done
by comparing same scenorio with Market AP's

Unit-Test : Attached
---
 src/ap/beacon.c            | 26 ++++++++++++++++++++++++--
 src/ap/ieee802_11.h        |  6 ++++--
 src/ap/ieee802_11_shared.c | 20 +++++++++++++++-----
 3 files changed, 43 insertions(+), 9 deletions(-)

diff --git a/src/ap/beacon.c b/src/ap/beacon.c
index c8de275eb..731f3adf4 100644
--- a/src/ap/beacon.c
+++ b/src/ap/beacon.c
@@ -753,6 +753,8 @@ struct probe_resp_params {
 #ifdef CONFIG_VENDOR_MXL
 	u8 *probe_req_mle;
 	size_t probe_req_mle_len;
+	const u8 *probe_req_ssid;
+	size_t probe_req_ssid_len;
 #endif /* CONFIG_VENDOR_MXL */
 };
 
@@ -886,6 +888,16 @@ static u8 * hostapd_probe_resp_fill_elems(struct hostapd_data *hapd,
 				size_t tx_bss_idx = mxl_hostapd_get_mbssid_tx_bss_idx(hapd->iface);
 				hapd = hapd->iface->bss[tx_bss_idx];
 			}
+			/* If Tx VAP & Non Tx VAP is configured hidden and Probe-Req is for Non-Tx VAP in Tx BSS */
+			/* then hide Tx VAP in probe response and make Non-Tx only visible*/
+			else if (os_memcmp(params->probe_req_ssid, hapd->conf->ssid.ssid, params->probe_req_ssid_len) != 0) {
+				if (params->probe_req_ssid_len != hapd->conf->ssid.ssid_len) {
+					is_req_non_transmit_vap = true;
+				}
+			}
+			else {
+				is_req_non_transmit_vap = false;
+			}
 			multiple_bssid_ie_len = multiple_bssid_ie_get_len(hapd, false, params->probe_req_mle,
 									  params->probe_req_mle_len);
 		}
@@ -954,7 +966,8 @@ static u8 * hostapd_probe_resp_fill_elems(struct hostapd_data *hapd,
 			return NULL;
 		}
 		pos = hostapd_eid_multiple_bssid(hapd, pos, false, params->probe_req_mle,
-						 params->probe_req_mle_len);
+						params->probe_req_mle_len, params->probe_req_ssid,
+						params->probe_req_ssid_len);
 		if (!pos) {
 			wpa_printf(MSG_ERROR, "hostapd_eid_multiple_bssid failed while probe resp template creation");
 			return NULL;
@@ -1922,6 +1935,9 @@ void handle_probe_req(struct hostapd_data *hapd,
 #ifdef CONFIG_VENDOR_MXL
 	params.is_ml_sta_info = false;
 	params.is_he_in_req = (elems.he_capabilities != NULL);
+	params.probe_req_ssid = elems.ssid;
+	params.probe_req_ssid_len = elems.ssid_len;
+
 #ifdef CONFIG_IEEE80211BE_MXL_MLO
 	params.probe_req_mle = (u8 *) elems.probe_req_mle;
 	params.probe_req_mle_len = elems.probe_req_mle_len;
@@ -1977,6 +1993,8 @@ void handle_probe_req(struct hostapd_data *hapd,
 #endif /* CONFIG_VENDOR_MXL */
 
 	os_free(params.resp);
+	params.probe_req_ssid = NULL;
+	params.probe_req_ssid_len = 0;
 
 	wpa_printf(MSG_EXCESSIVE, "STA " MACSTR " sent probe request for %s "
 		   "SSID", MAC2STR(mgmt->sa),
@@ -2026,6 +2044,8 @@ static u8 * hostapd_probe_resp_offloads(struct hostapd_data *hapd,
 	params.mld_ap = NULL;
 	params.mld_info = NULL;
 	params.is_he_in_req = 1;
+	params.probe_req_ssid = NULL;
+	params.probe_req_ssid_len = 0;
 
 	hostapd_gen_probe_resp(hapd, &params);
 	*resp_len = params.resp_len;
@@ -2077,6 +2097,8 @@ u8 * hostapd_unsol_bcast_probe_resp(struct hostapd_data *hapd,
 	probe_params.mld_ap = NULL;
 	probe_params.mld_info = NULL;
 	probe_params.is_he_in_req = 1;
+	probe_params.probe_req_ssid = NULL;
+	probe_params.probe_req_ssid_len = 0;
 
 	hostapd_gen_probe_resp(hapd, &probe_params);
 	ubpr->unsol_bcast_probe_resp_tmpl_len = probe_params.resp_len;
@@ -2623,7 +2645,7 @@ int ieee802_11_build_ap_params(struct hostapd_data *hapd,
 				"required length %d", multiple_bssid_ie_len);
 			goto fail;
 		}
-		tailpos = hostapd_eid_multiple_bssid(hapd, tailpos, true, NULL, 0);
+		tailpos = hostapd_eid_multiple_bssid(hapd, tailpos, true, NULL, 0, NULL, 0);
 		if (!tailpos) {
 			wpa_printf(MSG_ERROR, "hostapd_eid_multiple_bssid failed for beacon template creation");
 			goto fail;
diff --git a/src/ap/ieee802_11.h b/src/ap/ieee802_11.h
index 306b64853..a31d880b5 100644
--- a/src/ap/ieee802_11.h
+++ b/src/ap/ieee802_11.h
@@ -283,8 +283,10 @@ void notify_action_frame(struct hostapd_data *hapd, const u8 *action_frame,
 int multiple_bssid_ie_get_len(struct hostapd_data *hapd, bool is_beacon,
 			      u8 *probe_req_mle, size_t probe_req_mle_len);
 u8 * hostapd_eid_multiple_bssid(struct hostapd_data *hapd, u8 *eid, bool is_beacon,
-				u8 *probe_req_mle, size_t probe_req_mle_len);
-u8 * hostapd_eid_ssid(struct hostapd_data *hapd, u8 *eid);
+				u8 *probe_req_mle, size_t probe_req_mle_len,
+				const u8 *probe_req_ssid, size_t probe_req_ssid_len);
+u8 * hostapd_eid_ssid(struct hostapd_data *hapd, u8 *eid,
+				bool is_probe_req_ssid);
 
 #ifdef CONFIG_VENDOR_MXL
 u16 mxl_copy_supp_rates(struct hostapd_data *hapd, struct sta_info *sta,
diff --git a/src/ap/ieee802_11_shared.c b/src/ap/ieee802_11_shared.c
index 4fefec8f1..be7ef5483 100644
--- a/src/ap/ieee802_11_shared.c
+++ b/src/ap/ieee802_11_shared.c
@@ -1558,16 +1558,17 @@ int multiple_bssid_ie_get_len(struct hostapd_data *hapd, bool is_beacon,
 	return tot_len;
 }
 
-u8 * hostapd_eid_ssid(struct hostapd_data *hapd, u8 *eid)
+u8 * hostapd_eid_ssid(struct hostapd_data *hapd, u8 *eid, bool is_probe_req_ssid)
 {
 	errno_t err;
 	*eid++ = WLAN_EID_SSID;
-	if ( hapd->conf->ignore_broadcast_ssid == IGNORE_BROADCAST_SSID_CLEAR_SSID ) {
+
+	if ( (hapd->conf->ignore_broadcast_ssid == IGNORE_BROADCAST_SSID_CLEAR_SSID) && !is_probe_req_ssid) {
 		/* clear the data, but keep the correct length of the SSID */
 		*eid++ = hapd->conf->ssid.ssid_len;
 		os_memset(eid, 0, hapd->conf->ssid.ssid_len);
 		eid += hapd->conf->ssid.ssid_len;
-	} else if ( hapd->conf->ignore_broadcast_ssid ) {
+	} else if ( hapd->conf->ignore_broadcast_ssid && !is_probe_req_ssid ) {
 		*eid++ = 0; /* empty SSID */
 	} else {
 		*eid++ = hapd->conf->ssid.ssid_len;
@@ -1582,7 +1583,8 @@ u8 * hostapd_eid_ssid(struct hostapd_data *hapd, u8 *eid)
 }
 
 u8 * hostapd_eid_multiple_bssid(struct hostapd_data *hapd, u8 *eid, bool is_beacon,
-				u8 *probe_req_mle, size_t probe_req_mle_len)
+				u8 *probe_req_mle, size_t probe_req_mle_len,
+				const u8 *probe_req_ssid, size_t probe_req_ssid_len)
 {
 	int i;
 	u8 *data_pos, *size_pos, *subelem_data_pos, *subelem_size_pos;
@@ -1625,6 +1627,7 @@ u8 * hostapd_eid_multiple_bssid(struct hostapd_data *hapd, u8 *eid, bool is_beac
 	/* Iterate over each BSS */
 	for (i = MULTIBSS_REFERENCE_BSS_IDX; i < hapd->iface->num_bss; i++) {
 		struct hostapd_data *bss = hapd->iface->bss[i];
+		bool is_probe_req_ssid = false;
 
 		/* On BSS iteration skip TX BSS when building MBSSID IE */
 		if (mxl_hostapd_mbssid_is_tx_bss(bss->conf->bss_idx))
@@ -1656,8 +1659,15 @@ u8 * hostapd_eid_multiple_bssid(struct hostapd_data *hapd, u8 *eid, bool is_beac
 			wpa_printf(MSG_ERROR, "hostapd_eid_nontransmitted_bssid_capability failed");
 			goto end;
 		}
+		/* If Tx and Non-Tx VAP hidden, Probe Request carries Non-Tx SSID in Tx VAP */
+		/* In that case we have compare SSID to know its for Tx VAP or Non-Tx VAP */
+		if (!is_beacon && (probe_req_ssid_len == bss->conf->ssid.ssid_len)) {
+			if (os_memcmp(probe_req_ssid, bss->conf->ssid.ssid, probe_req_ssid_len) == 0) {
+				is_probe_req_ssid = true;
+			}
+		}
 		/* SSID element */
-		eid = hostapd_eid_ssid(bss, eid);
+		eid = hostapd_eid_ssid(bss, eid, is_probe_req_ssid);
 		if (!eid) {
 			wpa_printf(MSG_ERROR, "hostapd_eid_multiple_bssid ssid failed");
 			goto end;
-- 
2.43.0

